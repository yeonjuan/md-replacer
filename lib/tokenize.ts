import type { AnyToken, OtherToken, ReplaceToken } from "./types";

class Tokenizer {
  private static REPLACE_START_REGEX = /<!--\s*replace-start:([\s\S]*?)-->/;
  private static REPLACE_END_REGEX = /<!--\s*replace-end:([\s\S]*?)-->/;
  private index: number = 0;
  private tokens: AnyToken[] = [];
  private input: string = "";

  public tokenize(input: string): AnyToken[] {
    this.input = input;
    let remain: string = "";
    while ((remain = this.remain()).length) {
      let matched = null;
      let token: AnyToken | undefined;
      let end: number | undefined;
      if ((matched = remain.match(Tokenizer.REPLACE_START_REGEX))) {
        const [replaceStart, name] = matched;
        if (matched.index === 0) {
          token = this.finishReplaceToken(
            "replace-start",
            replaceStart,
            name.trim()
          );
        } else {
          end = matched.index;
        }
      } else if ((matched = remain.match(Tokenizer.REPLACE_END_REGEX))) {
        const [replaceEnd, name] = matched;
        if (matched.index === 0) {
          token = this.finishReplaceToken(
            "replace-end",
            replaceEnd,
            name.trim()
          );
        } else {
          end = matched.index;
        }
      }
      if (!token) {
        token = this.finishOtherToken(this.remain().slice(0, end));
      }
      this.pushToken(token);
    }
    return this.tokens;
  }

  private pushToken(token: AnyToken) {
    this.tokens.push(token);
  }

  private finishOtherToken(raw: string): OtherToken {
    const start = this.cursor();
    this.eatText(raw);
    const end = this.cursor();
    const token: OtherToken = {
      type: "other",
      raw,
      pos: {
        start,
        end,
      },
    };
    return token;
  }

  private finishReplaceToken(
    type: ReplaceToken["type"],
    raw: string,
    name: string
  ): ReplaceToken {
    const start = this.cursor();
    this.eatText(raw);
    const end = this.cursor();
    const token: ReplaceToken = {
      type,
      raw,
      name,
      pos: {
        start: start,
        end: end,
      },
    };
    return token;
  }

  private cursor(): number {
    return this.index;
  }

  private eatText(text: string): void {
    this.index += text.length;
    this.input = this.input.slice(text.length);
  }

  private remain(): string {
    return this.input;
  }
}

function tokenize(input: string): AnyToken[] {
  return new Tokenizer().tokenize(input);
}
export = tokenize;
