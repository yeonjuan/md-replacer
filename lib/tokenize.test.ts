import tokenize = require("./tokenize");

describe("tokenize", () => {
  it("tokenize - no replacer", () => {
    const markdown = `
content
`;
    expect(tokenize(markdown)).toMatchSnapshot();
  });

  it("tokenize - one replacer", () => {
    const markdown = `
content 1
<!-- start: foo -->
content 2
<!-- end: foo -->
`;
    expect(tokenize(markdown)).toMatchSnapshot();
  });

  it("tokenize - two replacer", () => {
    const markdown = `
content 1
<!-- start: foo -->
content 2
<!-- end: foo -->
content 3
<!-- start: bar -->
content 4
<!-- end: bar -->
content 5
`;
    expect(tokenize(markdown)).toMatchSnapshot();
  });

  it("tokenize - two replace - same line", () => {
    const markdown = `
before
<!-- start: foo -->
foo
<!-- end: foo -->between<!-- start: bar -->bar<!-- end: bar -->after`;
    expect(tokenize(markdown)).toMatchSnapshot();
  });
});
