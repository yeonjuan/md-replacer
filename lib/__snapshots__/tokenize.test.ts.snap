// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenize tokenize - no replacer 1`] = `
Array [
  OtherToken {
    "pos": Object {
      "end": 9,
      "start": 0,
    },
    "raw": "
content
",
    "type": "other",
  },
]
`;

exports[`tokenize tokenize - one replacer 1`] = `
Array [
  OtherToken {
    "pos": Object {
      "end": 11,
      "start": 0,
    },
    "raw": "
content 1
",
    "type": "other",
  },
  ReplaceStartToken {
    "name": "foo",
    "pos": Object {
      "end": 38,
      "start": 11,
    },
    "raw": "<!-- replace-start: foo -->",
    "type": "replace-start",
  },
  OtherToken {
    "pos": Object {
      "end": 49,
      "start": 38,
    },
    "raw": "
content 2
",
    "type": "other",
  },
  ReplaceEndToken {
    "name": "foo",
    "pos": Object {
      "end": 74,
      "start": 49,
    },
    "raw": "<!-- replace-end: foo -->",
    "type": "replace-end",
  },
  OtherToken {
    "pos": Object {
      "end": 75,
      "start": 74,
    },
    "raw": "
",
    "type": "other",
  },
]
`;

exports[`tokenize tokenize - two replace - same line 1`] = `
Array [
  OtherToken {
    "pos": Object {
      "end": 8,
      "start": 0,
    },
    "raw": "
before
",
    "type": "other",
  },
  ReplaceStartToken {
    "name": "foo",
    "pos": Object {
      "end": 35,
      "start": 8,
    },
    "raw": "<!-- replace-start: foo -->",
    "type": "replace-start",
  },
  OtherToken {
    "pos": Object {
      "end": 40,
      "start": 35,
    },
    "raw": "
foo
",
    "type": "other",
  },
  ReplaceEndToken {
    "name": "foo",
    "pos": Object {
      "end": 65,
      "start": 40,
    },
    "raw": "<!-- replace-end: foo -->",
    "type": "replace-end",
  },
  OtherToken {
    "pos": Object {
      "end": 72,
      "start": 65,
    },
    "raw": "between",
    "type": "other",
  },
  ReplaceStartToken {
    "name": "bar",
    "pos": Object {
      "end": 99,
      "start": 72,
    },
    "raw": "<!-- replace-start: bar -->",
    "type": "replace-start",
  },
  OtherToken {
    "pos": Object {
      "end": 102,
      "start": 99,
    },
    "raw": "bar",
    "type": "other",
  },
  ReplaceEndToken {
    "name": "bar",
    "pos": Object {
      "end": 127,
      "start": 102,
    },
    "raw": "<!-- replace-end: bar -->",
    "type": "replace-end",
  },
  OtherToken {
    "pos": Object {
      "end": 132,
      "start": 127,
    },
    "raw": "after",
    "type": "other",
  },
]
`;

exports[`tokenize tokenize - two replacer 1`] = `
Array [
  OtherToken {
    "pos": Object {
      "end": 11,
      "start": 0,
    },
    "raw": "
content 1
",
    "type": "other",
  },
  ReplaceStartToken {
    "name": "foo",
    "pos": Object {
      "end": 38,
      "start": 11,
    },
    "raw": "<!-- replace-start: foo -->",
    "type": "replace-start",
  },
  OtherToken {
    "pos": Object {
      "end": 49,
      "start": 38,
    },
    "raw": "
content 2
",
    "type": "other",
  },
  ReplaceEndToken {
    "name": "foo",
    "pos": Object {
      "end": 74,
      "start": 49,
    },
    "raw": "<!-- replace-end: foo -->",
    "type": "replace-end",
  },
  OtherToken {
    "pos": Object {
      "end": 85,
      "start": 74,
    },
    "raw": "
content 3
",
    "type": "other",
  },
  ReplaceStartToken {
    "name": "bar",
    "pos": Object {
      "end": 112,
      "start": 85,
    },
    "raw": "<!-- replace-start: bar -->",
    "type": "replace-start",
  },
  OtherToken {
    "pos": Object {
      "end": 123,
      "start": 112,
    },
    "raw": "
content 4
",
    "type": "other",
  },
  ReplaceEndToken {
    "name": "bar",
    "pos": Object {
      "end": 148,
      "start": 123,
    },
    "raw": "<!-- replace-end: bar -->",
    "type": "replace-end",
  },
  OtherToken {
    "pos": Object {
      "end": 159,
      "start": 148,
    },
    "raw": "
content 5
",
    "type": "other",
  },
]
`;
